### PR TITLE
added support for proxy server and fixed bug #23 and #21

### PR DESCRIPTION
**Pull request by rocklan**
This pull request adds support for specifying a proxy server when making the request to Slack's API. It also fixes the bug where properties are not serialised as JSON. Added a few unit tests too and a few other misc things.

### CODE REVIEW

File: src/Seq.App.Slack/Seq.App.Slack.nuspec [hunk 1]
```diff
@@ -6,7 +6,7 @@
     <requireLicenseAcceptance>false</requireLicenseAcceptance>
     <description>An app for Seq (https://getseq.net) that forwards events to Slack.</description>
     <language>en-US</language>
-    <projectUrl>https://github.com/bytenik/Seq.App.Slack</projectUrl>
+    <projectUrl>https://github.com/rocklan/Seq.App.Slack</projectUrl>
```
Comment (rocklan) [inline src/Seq.App.Slack/Seq.App.Slack.nuspec]: oops feel free to set this back :)

File: src/Seq.App.Slack/SlackApi.cs [hunk 2]
```diff
@@ -1,24 +1,48 @@
...
 using System.Text;
 using Newtonsoft.Json;
 
 namespace Seq.App.Slack
 {
-    static class SlackApi
+    class SlackApi
     {
-        private static readonly HttpClient HttpClient = new HttpClient();
+        private HttpClient httpClient;
```
Comment (nblumhardt) [inline src/Seq.App.Slack/SlackApi.cs]: `readonly`?

File: src/Seq.App.Slack/SlackApi.cs [hunk 3]
```diff
@@ -1,24 +1,48 @@
...
 
-        private static readonly JsonSerializerSettings JsonSerializerSettings = new JsonSerializerSettings
```
Comment (nblumhardt) [inline src/Seq.App.Slack/SlackApi.cs]: Any reason for the change in serializer settings scope here?

File: src/Seq.App.Slack/SlackReactor.cs [hunk 4]
```diff
@@ -59,10 +55,23 @@ public class SlackReactor : Reactor, ISubscribeTo<LogEventData>
...
+        [SeqAppSetting(
+            DisplayName = "Proxy Server",
+            HelpText = "Proxy server to be used when making HTTPS request to slack api, uses default credentials",
+            IsOptional = true)]
+        public string ProxyServer { get; set; }
+
+        [SeqAppSetting(
+            DisplayName = "Dictionary serialiser string limit",
+            IsOptional = true,
+            HelpText = "If a property is a Dictionary, it's serialised as JSON and will be truncated if the resulting string is longer than this number")]
```
Comment (nblumhardt) [inline src/Seq.App.Slack/SlackReactor.cs]: Could this just apply to all property values? I can see arrays suffering from similar issues.

File: src/Seq.App.Slack/SlackReactor.cs [hunk 5]
```diff
@@ -59,10 +55,23 @@ public class SlackReactor : Reactor, ISubscribeTo<LogEventData>
...
+        public int? JsonTrunateAt { get; set; } = null;
+
         private EventTypeSuppressions _suppressions;
 
         private static readonly IImmutableList<string> SpecialProperties = ImmutableList.Create("Id", "Host");
 
+        private static SlackApi slackApi;
```
Comment (nblumhardt) [inline src/Seq.App.Slack/SlackReactor.cs]: Nit, whitespace, and `_propertyName` convention

File: src/Seq.App.Slack/SlackApi.cs (hunk 2 above)
Comment (nblumhardt) [reply src/Seq.App.Slack/SlackApi.cs]: Nit: the codebase uses `_underscore` property name prefixes.

File: src/Seq.App.Slack/SlackReactor.cs [hunk 6]
```diff
@@ -129,19 +141,51 @@ public void On(Event<LogEventData> evt)
...
             }
 
             if (otherProperties.Fields.Count != 0)
                 message.Attachments.Add(otherProperties);
 
-            SlackApi.SendMessage(WebhookUrl, message);
+            slackApi.SendMessage(WebhookUrl, message);
+        }
+
+        internal string convertPropertyValueToString(object propertyValue)
```
Comment (nblumhardt) [inline src/Seq.App.Slack/SlackReactor.cs]: Nit - codebase uses `CamelCase()` here

File: src/Seq.App.Slack/Properties/AssemblyInfo.cs [hunk 7]
```diff
@@ -1,3 +1,17 @@
...
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
 
 [assembly: InternalsVisibleTo("Seq.App.Slack.Tests")]
+[assembly: AssemblyTitle("seq.app.slack")]
+[assembly: AssemblyDescription("An app for Seq (http://getseq.net) that forwards messages to Slack.")]
+[assembly: AssemblyConfiguration("")]
+[assembly: AssemblyCompany("Datalust")]
+[assembly: AssemblyProduct("seq.app.slack")]
+[assembly: AssemblyCopyright("Copyright © Datalust 2018")]
```
[the remark this answers is not in the document]
Comment (rocklan) [reply src/Seq.App.Slack/Properties/AssemblyInfo.cs]: I've removed Datalust from company and copyright

File: src/Seq.App.Slack/SlackApi.cs (hunk 2 above)
Comment (rocklan) [reply src/Seq.App.Slack/SlackApi.cs]: fixed, made readonly and used underscore

File: src/Seq.App.Slack/SlackApi.cs (hunk 3 above)
Comment (rocklan) [reply src/Seq.App.Slack/SlackApi.cs]: no good reason, I've reverted it

File: src/Seq.App.Slack/SlackReactor.cs (hunk 4 above)
Comment (rocklan) [reply src/Seq.App.Slack/SlackReactor.cs]: Makes sense, it applies to everything now. Would it make sense to set it to a sensible limit (500?) by default? I've just left it as null by default

File: src/Seq.App.Slack/SlackReactor.cs (hunk 5 above)
Comment (rocklan) [reply src/Seq.App.Slack/SlackReactor.cs]: fixed

File: src/Seq.App.Slack/SlackReactor.cs (hunk 6 above)
Comment (rocklan) [reply src/Seq.App.Slack/SlackReactor.cs]: fixed

File: src/Seq.App.Slack/SlackReactor.cs (hunk 4 above)
Comment (nblumhardt) [reply src/Seq.App.Slack/SlackReactor.cs]: Awesome, thanks for all the updates :+1: 

I think defaulting to `null` is probably the most conservative choice for a start, perhaps we see how this gets used, and consider another default down the track.

**Pull request merged**